# Import modules

In [1]:
import pandas as pd

# II. Exploring data

In [3]:
file = pd.read_csv('MELBOURNE_HOUSE_PRICES_LESS.csv')
file.head()

Suburb           Address  Rooms Type      Price Method   SellerG  \
0    Abbotsford     49 Lithgow St      3    h  1490000.0      S    Jellis   
1    Abbotsford     59A Turner St      3    h  1220000.0      S  Marshall   
2    Abbotsford     119B Yarra St      3    h  1420000.0      S    Nelson   
3    Aberfeldie        68 Vida St      3    h  1515000.0      S     Barry   
4  Airport West  92 Clydesdale Rd      2    h   670000.0      S    Nelson   

        Date  Postcode             Regionname  Propertycount  Distance  \
0  1/04/2017      3067  Northern Metropolitan           4019       3.0   
1  1/04/2017      3067  Northern Metropolitan           4019       3.0   
2  1/04/2017      3067  Northern Metropolitan           4019       3.0   
3  1/04/2017      3040   Western Metropolitan           1543       7.5   
4  1/04/2017      3042   Western Metropolitan           3464      10.4   

                  CouncilArea  
0          Yarra City Council  
1          Yarra City Council  
2          Yarra City Council  
3  Moonee Valley City Council  
4  Moonee Valley City Council

### Kiểm tra xem có bao nhiều dòng và cột trong dữ liệu | dataset

In [4]:
num_rows, num_cols = file.shape
file.shape

(63023, 13)

### Ý nghĩa của từng dòng trong dữ liệu | dataset

Theo dữ liệu mô tả và yêu cầu của tác giả của chủ đề "Melbourne Housing Market". Melbourne ở Úc đã trải qua hiện tượng bong bóng nhà đất (một số chuyên gia cho rằng nó có thể vỡ sớm) trong năm 2016 - 2017.

    Và theo sự quan sát sơ bộ về tập dữ liệu thì một đòng cho biết thông tin của một ngôi nhà đang được rao giá, và hầu hết dữ liệu đều như vậy.

### Kiểm tra xem có dòng nào bị trùng trong bộ dữ liệu không ?

In [5]:
have_duplicated_rows = bool
dup_df = file[file.duplicated(keep=False)]
if len(dup_df) == 0:
    have_duplicated_rows = False
else:
    have_duplicated_rows = True
print(have_duplicated_rows)
dup_df

True


Suburb         Address  Rooms Type      Price Method   SellerG  \
8041   Fitzroy North  5/16 Taplin St      2    h  1010000.0     SP  Woodards   
8042   Fitzroy North  5/16 Taplin St      2    h  1010000.0     SP  Woodards   
33941     Nunawading   1/7 Lilian St      3    t        NaN     SP    Jellis   
33942     Nunawading   1/7 Lilian St      3    t        NaN     SP    Jellis   

             Date  Postcode             Regionname  Propertycount  Distance  \
8041    5/05/2018      3068  Northern Metropolitan           6244       3.6   
8042    5/05/2018      3068  Northern Metropolitan           6244       3.6   
33941  17/06/2017      3131   Eastern Metropolitan           4973      15.4   
33942  17/06/2017      3131   Eastern Metropolitan           4973      15.4   

                   CouncilArea  
8041     Moreland City Council  
8042     Moreland City Council  
33941  Manningham City Council  
33942  Manningham City Council

Từ kết quả trên ta có thể dễ nhận thấy dữ liệu bị trùng, nối tiếp nhau. Dòng 8042 lặp lại của dòng 8041 và 33942 lặp lại của 33941.

    Kết luận có lẽ vào thời điểm đó tác giả ngừng thu thập dữ liệu và ngày tiếp theo vẫn tiếp tục lưu trữ bị trùng ở mặt đó

### Mỗi cột sẽ có ý nghĩa như thế nào ?

Mô tả về các cột | attributes của tập dữ liệu "MELBOURNE_HOUSE_PRICES_LESS.csv".
- **Suburb**: ngoại ô, nơi không nằm trong khu vực của Melbourne CBD.
- **Address**: địa chỉ của ngôi nhà.
- **Rooms**: số lượng phòng có trong ngôi nhà.
- **SellerG**: đại lí bất động sản
- **Date**: ngày được bán
- **Regionname**: khu vực trong Melbourne CBD
- **Propertycount**: số lượng bất động sản tồn tại ở vùng ngoại ô (Suburb)
- **Distance**: khoảng cách kilometers từ ngôi nhà so với Melbourne CBD (trung tâm thương mai Melbourne) 
- **CouncilArea** Hội động quản trị khu vực (Governing council for the area)
- **PostCode**: mã bưu chính
- **Price**: giá ngôi nhà
<br> </br>
- **Type**: phân loại của ngôi nhà
    - Nếu là **br** - bedroom(s);
    - Nếu là **h** - house,cottage,villa, semi,terrace;
    - Nếu là **u** - unit, duplex;
    - Nếu là **t** - townhouse;
    - Nếu là **dev site** - development site;
    - Nếu là **o res** - other residential.  
- **Method**: phương thức giao dịch bất động sản
    - Nếu là **S** - property sold;
    - Nếu là **SP** - property sold prior;
    - Nếu là **PI** - property passed in;
    - Nếu là **PN** - sold prior not disclosed;
    - Nếu là **SN** - sold not disclosed;
    - Nếu là **NB** - no bid;
    - Nếu là **VB** - vendor bid;
    - Nếu là **W** - withdrawn prior to auction;
    - Nếu là **SA** - sold after auction;
    - Nếu là **SS** - sold after auction price not disclosed.
    - Nếu là **N/A** - price or highest bid not available.

### Kiểm tra từng kiểu dữ liệu của từng cột. Có cột nào có dữ liệu chưa phù hợp hay không ?

In [6]:
file.dtypes

Suburb            object
Address           object
Rooms              int64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Postcode           int64
Regionname        object
Propertycount      int64
Distance         float64
CouncilArea       object
dtype: object

Hiện tại có dữ liệu của cột <b>Date</b> chưa đúng với dạng "daytime", xử lí dữ liệu phần dưới !

In [7]:
file['Date'] = pd.to_datetime(file['Date'])
file.head()

Suburb           Address  Rooms Type      Price Method   SellerG  \
0    Abbotsford     49 Lithgow St      3    h  1490000.0      S    Jellis   
1    Abbotsford     59A Turner St      3    h  1220000.0      S  Marshall   
2    Abbotsford     119B Yarra St      3    h  1420000.0      S    Nelson   
3    Aberfeldie        68 Vida St      3    h  1515000.0      S     Barry   
4  Airport West  92 Clydesdale Rd      2    h   670000.0      S    Nelson   

        Date  Postcode             Regionname  Propertycount  Distance  \
0 2017-01-04      3067  Northern Metropolitan           4019       3.0   
1 2017-01-04      3067  Northern Metropolitan           4019       3.0   
2 2017-01-04      3067  Northern Metropolitan           4019       3.0   
3 2017-01-04      3040   Western Metropolitan           1543       7.5   
4 2017-01-04      3042   Western Metropolitan           3464      10.4   

                  CouncilArea  
0          Yarra City Council  
1          Yarra City Council  
2          Yarra City Council  
3  Moonee Valley City Council  
4  Moonee Valley City Council

# Với mỗi cột có kiểu dữ liệu dạng numeric, các giá trị được phân bố như thế nào?
Các cột numeric gồm có "Price", "Date", "Distance"

In [8]:
numeric_df = file[["Price", "Date", "Distance"]]
percent_missing = numeric_df.isnull().sum() * 100 / len(numeric_df)
percent_missing = round(percent_missing, 2)
min_df = numeric_df.min(axis = 0)
max_df = numeric_df.max(axis = 0)
nume_col_profiles_df = pd.DataFrame([percent_missing, min_df, max_df], index=['missing_ratio', 'min', 'max'])
nume_col_profiles_df

Price                 Date  Distance
missing_ratio        23.15                  0.0       0.0
min               85000.00  2016-01-28 00:00:00       0.0
max            11200000.00  2018-12-05 00:00:00      64.1

# Với mỗi cột có kiểu dữ liệu dạng categorical, các giá trị được phân bố như thế nào?
Các côt categorical gồm có "Suburb", "Address", "Rooms", "Type", "Method", "SellerG", "Postcode", "Regionname", "Propertycount", "CouncilArea"

In [10]:
categorical_df = file[["Suburb", "Address", "Rooms", "Type", "Method", "SellerG", "Postcode", "Regionname", "Propertycount", "CouncilArea"]]
percent_missing = categorical_df.isnull().sum() * 100 / len(numeric_df)
percent_missing = round(percent_missing, 2)
num_unique_values = categorical_df.nunique()
lst = []
for col in categorical_df:
    lst.append(categorical_df[col].dropna().unique())
unique_values = pd.Series(lst, index=["Suburb", "Address", "Rooms", "Type", "Method", "SellerG", "Postcode", "Regionname", "Propertycount", "CouncilArea"])
cate_col_profiles_df = pd.DataFrame([percent_missing, num_unique_values, unique_values], index=['missing_ratio', 'num_diff_vals', 'diff_vals'])
cate_col_profiles_df

Suburb  \
missing_ratio                                                0.0   
num_diff_vals                                                380   
diff_vals      [Abbotsford, Aberfeldie, Airport West, Albanva...   

                                                         Address  \
missing_ratio                                                0.0   
num_diff_vals                                              57754   
diff_vals      [49 Lithgow St, 59A Turner St, 119B Yarra St, ...   

                                                         Rooms       Type  \
missing_ratio                                              0.0        0.0   
num_diff_vals                                               14          3   
diff_vals      [3, 2, 6, 4, 1, 5, 7, 8, 11, 10, 12, 9, 31, 16]  [h, t, u]   

                                           Method  \
missing_ratio                                 0.0   
num_diff_vals                                   9   
diff_vals      [S, SP, PI, SN, VB, PN, SA, W, SS]   

                                                         SellerG  \
missing_ratio                                                0.0   
num_diff_vals                                                476   
diff_vals      [Jellis, Marshall, Nelson, Barry, hockingstuar...   

                                                        Postcode  \
missing_ratio                                                0.0   
num_diff_vals                                                225   
diff_vals      [3067, 3040, 3042, 3021, 3206, 3020, 3078, 301...   

                                                      Regionname  \
missing_ratio                                                0.0   
num_diff_vals                                                  8   
diff_vals      [Northern Metropolitan, Western Metropolitan, ...   

                                                   Propertycount  \
missing_ratio                                                0.0   
num_diff_vals                                                368   
diff_vals      [4019, 1543, 3464, 1899, 3280, 2185, 2211, 530...   

                                                     CouncilArea  
missing_ratio                                                0.0  
num_diff_vals                                                 34  
diff_vals      [Yarra City Council, Moonee Valley City Counci...